### PR TITLE
[feat] Send HTTP GET header with settings when looking for a 'package_reference'

### LINKED ISSUE #7870: [feature] Identify configuration in remote requests when querying for a packageID
When Conan checks if a package for a given configuration is available in a remote, it sends just the _package ID_. It would be very useful to send the information used to compute that hash:
 * We can send _all_ the configuration: _package ID_ computation is reproducible (needed? desired?)
 * We can send only the information taken into account to compute the _package ID_ (~`conaninfo.txt`)

### PR DESCRIPTION
**Pull request by jgsogo**
Changelog: Feature: Add headers with settings and options to HTTP GET requests when searching for packages.
Docs: Omit

closes https://github.com/conan-io/conan/issues/7870

Conan sends a request header `Conan-PkgID-Settings` when looking for a `pref` in the remotes:
 * if using revisions, it is sent when looking for the latest package revision, URL like:
   ```
   v2/conans/<reference>/revisions/<rrev>/packages/<package_id>/latest
   ```

 * if not using revisions, it is in the downloads_url:
   ```
   v1/conans/<reference>/packages/<package_id>/download_urls
   ```

These are the headers, do we want this information in a single header? do we want to use different names?

```
Conan-PkgID-Settings
Conan-PkgID-Options
```

---

#REVISIONS: 1

### CODE REVIEW

File: conans/test/functional/remote/test_request_headers.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+        self.assertNotIn('build_type', settings_header)
+
+    def _assert_options_headers(self, options_header, shared_value='False'):
+        self.assertListEqual(['shared'], [it.split('=', 1)[0] for it in options_header.split(';')])
+        self.assertIn('shared={}'.format(shared_value), options_header)
+
+    def _get_test_client(self):
+        t = TestClient(requester_class=RequesterClass, servers=self.servers,
+                       users={"default": [("user", "mypass")]})
+        t.run('config set general.revisions_enabled={}'.format('1' if self.revs_enabled else '0'))
```
Comment (memsharded) [inline conans/test/functional/remote/test_request_headers.py:83]: This is running the same test 2 times, right? Once with TESTING_REVISIONS_ENABLED and once without it, then overriding the testing one. I would suggest using the ``cache.config.revisions_enabled`` to check what is the current value defined by testing.

File: conans/client/graph/graph_binaries.py [hunk 2]
```diff
@@ -107,9 +110,9 @@ def _evaluate_remote_pkg(self, node, pref, remote, remotes):
         # If the "remote" came from the registry but the user didn't specified the -r, with
         # revisions iterate all remotes
         if not remote or (not remote_info and self._cache.config.revisions_enabled):
-            for r in remotes.values():
+            for r in remotes.values():  # FIXME: Here we hit the same remote we did before
```
Comment (memsharded) [inline conans/client/graph/graph_binaries.py:113]: good catch

File: conans/client/conan_api.py [hunk 3]
```diff
@@ -1133,7 +1133,8 @@ def get_path(self, reference, package_id=None, path=None, remote_name=None):
             if package_id:
                 pref = PackageReference(ref, package_id)
                 if self.app.config.revisions_enabled and not pref.revision:
-                    pref = self.app.remote_manager.get_latest_package_revision(pref, remote)
+                    pref = self.app.remote_manager.get_latest_package_revision(pref, remote,
```
Comment (memsharded) [inline conans/client/conan_api.py]: Maybe define a default=None for headers arg?

**Review by memsharded — APPROVED** [review]
Looks good

**Review by czoido — APPROVED** [review]

**Review by memsharded — APPROVED** [review]
Please make sure that it won't break different servers like Artifactory, Bintray, etc.

**Comment by jgsogo** [comment]
I've tried a local Artifactory in docker, my Bintray account, and a cloud Artifactory (staging environment for ConanCenter).

**Pull request merged**